# **HOMEWORK 3...?** 🕵
## **DUE: *MARCH 25, 2025 @ 11:59 PM***
## **24-HR LATE DUE DATE W/ A 15% PENALTY: *MARCH 26, 2025 @ 11:59 PM***

### **DO NOT REMOVE ANY PART OF ANY OF THE QUESTIONS OR YOU LOSE CREDIT**
### *No Hardcoding Any Part of Any Question*  😋🪄🧌
### **REMEMBER TO SHOW ALL CODE OUTPUT (NO CREDIT OTHERWISE)**
### **Reminder: Please make sure your code runs before submitting your work. Code sections that do not run will receive 0 credits, no partials will be given. This is VERY important in real project development.**

---
---
**DISCLAIMER: This is a fake story, not a real incident! We have created this story just for the data exploration purpose.**

---
---

**Campus Police Department**

Incident Report

---

**Incident Date:** Febuary 27, 2025

**Location of Incident:** 2nd Floor, Iribe Building

**Items Left at the Scene:** Red Ticket for Zichao's Magic Show

**Case Number:** 2025-IRB-0027

---
**Victim Information:**

**Name:** Dr. Fardina Alam

**Affiliation:** University of Maryland, College Park

**Position:** Professor

**Contact Information:** fardina@umd.edu

---

**Incident Description:**

On the evening of Febuary 27, 2025, an unknown individual(s) unlawfully entered Dr. Fardina's room located on the 2nd floor of Iribe and stole a set of Markers. The stolen items were a set of limited edition markers that Dr. Fardina had spent years collecting. At the scene of the crime, a red entry ticket for "Zichao's Magic Show" was found on the floor near Dr. Fardina's desk.






In [3]:
# We import all essential libraries for this HW.
# If you want to add additional libraries, you can add more here!
import pandas as pd
from datetime import datetime
import string
from collections import Counter
import statistics
from scipy.stats import mode

## YOUR TASK 🔎:

In this assignment, you are tasked with solving the mystery of Dr. Fardina's stolen markers and finding the culprit.

* To accomplish this, you will be provided with **4 data sets** containing key information related to the incident. The list of the data sets name given below:

 - Profiles
 - Witnesses
 - Zichao's magic show
 - Interrogation Statements



* Your goal is to analyze these data sets to uncover the clues and evidences that will lead you to the thief.

**Be careful in following instructions as it may lead to you losing points :(**

Take a moment to look at all the data sets below:

---

**Profiles:**

This dataset contains information regarding all the potential people that could have committed the crime. General information regarding each person are as follows:
- **ID:** A **unique** identification for each person
- **Name:** The name of the person
- **Birthday:** The birthday of the person
- **Location_during_crime:** The location of the person during the crime
- **Eye_Color:** The eye color of the person
- **Skin_Color:** The skin color of the person
- **Hair_Color:** The hair color of the person
- **Height:** The height of the person in cm

**Witnesses:**

This dataset contains a statement from everyone in profiles, regarding suspicious things they saw during the time of the crime. Information that comes in this dataset are as follows:
- **ID:** A **Unique** identification for each person
- **Statement:** A statement from the person regarding a suspicious activity during the time of the crime

**Zichao's Magic Show:**

This dataset contains everyone that purchased a ticket for the show. **NOTE** that not everyone in profiles has purchased a ticket for this show. The Information contained in this dataset are as follows:
- **ID:** A **Unique** indentification for each person
- **Ticket_Number:** A **Unique** number that serves as an identification for the ticket bought

**Interrogation Statements:**

This datasets contains a statement from each person after they have been interrogated. **NOTE** assume that each statement is ***100% true*** and that ***there are people in this dataset who aren't in the "Profiles" dataset***.
Information contained in this dataset are as follows:

- **Name:** The name of the person being interrogated
- **Statement:** A statement the person gives after being interrogated

---

## Q1) Lets begin by observing the profiles dataset (10 PTS Total)






**TASK 1.1 (1 Point):**  In the code box below, load in "Profiles" dataset into a variable called **profiles_df** and display it

In [4]:
profiles_df = pd.read_csv("Profiles.csv")
profiles_df

ID              Name    Birthday Location_during_crime Eye_Color  \
0    P299       Marco Rossi  1969/07/10          fourth floor       red   
1    P086       Aarav Patel  1979/03/10          second floor    indigo   
2    P022            Li Wei  1990/10/08          fourth floor      blue   
3    P003     Haruto Tanaka  1996/27/10                 lobby     green   
4    P036      Kwame Mensah  1991/26/02          fourth floor      blue   
..    ...               ...         ...                   ...       ...   
343  P126       Andrew Choe  1988/03/01                   NaN    violet   
344  P203     Krish Thakker  1965/26/09           third floor    yellow   
345  P273     Jeonghan Yoon  2003/12/07          fourth floor      blue   
346  P290  Ananya Malipeddi  10/03/1994           third floor       NaN   
347  P233         Ifra Syed  06/27/1982          second floor       red   

    Skin_Color Hair_Color  Height  
0       indigo     orange     191  
1       indigo      green     198  
2          red     violet     154  
3         blue     violet     158  
4       yellow     orange     184  
..         ...        ...     ...  
343     indigo        red     176  
344     indigo     orange     193  
345     indigo        red     193  
346     yellow     indigo     161  
347     yellow     orange     169  

[348 rows x 8 columns]

**TASK 1.2 (1 Point):**  Next, you need to display the number of rows and columns in the "Profiles" dataset.

In [5]:
profiles_df.shape

(348, 8)

**TASK 1.3 (1 Point):**  Next, you need to display the name of the columns in the "Profiles" dataset.

In [6]:
list(profiles_df)

['ID',
 'Name',
 'Birthday',
 'Location_during_crime',
 'Eye_Color',
 'Skin_Color',
 'Hair_Color',
 'Height']

**TASK 1.4 (1 Point):**  Now, you need to display the types of the columns in the "Profiles" dataset.

In [7]:
profiles_df.dtypes

ID                       object
Name                     object
Birthday                 object
Location_during_crime    object
Eye_Color                object
Skin_Color               object
Hair_Color               object
Height                    int64
dtype: object

**TASK 1.5 (2 Points):** Utilizing the `count()` function on this dataset and set it to the variable **info** and then display it.

Write any **potential problem** that you see with the dataset when observing the output of the count function:




In [8]:
info = profiles_df.count() # opposite is: profiles_df.isnull().sum()
print(info)

ID                       348
Name                     348
Birthday                 348
Location_during_crime    280
Eye_Color                280
Skin_Color               348
Hair_Color               348
Height                   348
dtype: int64


**<font color='red'>Write your observations here:</font>**

The Eye_Color and Location_during_crime columns only have 280 non-null values, which means there are 68 null values in each column. This could mess with our investigation.

**TASK 1.6 (4 Points):**

Look back at the "Profiles" dataset. Using other pandas functions than `count()`, **can you spot another potential problem with the dataset?**

In [64]:
profiles_df['ID'].value_counts()

ID
P299    2
P171    2
P294    2
P298    2
P300    2
       ..
P011    1
P093    1
P185    1
P030    1
P233    1
Name: count, Length: 300, dtype: int64

**<font color='red'>Write your observations here:</font>**

Some of the IDs have duplicates. Which I can run into issues later when looking at interrogations and witnesses.

## Q2) Cleaning the Dataset (25 PTS Total)

**TASK 2.1 (10 Points):** Lets begin by fixing the Birthday column. You may notice that the dates are in inconsistent format!

- Pandas internally represent dates in 'datetime64' format which is a form of ISO 8601. Basically, dates are represented as yyyy-mm-dd.


- Fix all the dates in the Birthday Column.
- Display the "Profiles" dataframe after you are done.
- Note that the only **2 formats** the dates are in is either **yyyy/dd/mm** or **mm/dd/yyyy**;
- You want to represent them as **yyyy-mm-dd**

- **EXAMPLE**: In the “Profiles” dataset, dates like *2000/10/12* and *10/12/2000* follow different formats—one as *December 10th, 2000*, and the other as *October 12th, 2000*. We aim to standardize them to *2000-12-10* and *2000-10-12*, respectively.

**HINT:** some datetime functions might make things easier.
- https://www.geeksforgeeks.org/python-pandas-to_datetime/
- https://www.programiz.com/python-programming/datetime/strftime


In [10]:
def fixFormat(birthday):
    date_split = birthday.split("/")
    
    # all need to be in the format-> "year-month-day"
    if int(date_split[0]) < 13: # means the first digit is a month
        # month/day/year
        modified_birthday = date_split[2] + "-" + date_split[0] + "-" + date_split[1]
        
    else: # first digit is year
        #year/day/month
        modified_birthday = date_split[0] + "-" + date_split[2] + "-" + date_split[1]
        
    # convert to date_time
    date = pd.to_datetime(modified_birthday)
    return date

    
profiles_df['Birthday'] = profiles_df['Birthday'].apply(fixFormat)
profiles_df


ID              Name   Birthday Location_during_crime Eye_Color  \
0    P299       Marco Rossi 1969-10-07          fourth floor       red   
1    P086       Aarav Patel 1979-10-03          second floor    indigo   
2    P022            Li Wei 1990-08-10          fourth floor      blue   
3    P003     Haruto Tanaka 1996-10-27                 lobby     green   
4    P036      Kwame Mensah 1991-02-26          fourth floor      blue   
..    ...               ...        ...                   ...       ...   
343  P126       Andrew Choe 1988-01-03                   NaN    violet   
344  P203     Krish Thakker 1965-09-26           third floor    yellow   
345  P273     Jeonghan Yoon 2003-07-12          fourth floor      blue   
346  P290  Ananya Malipeddi 1994-10-03           third floor       NaN   
347  P233         Ifra Syed 1982-06-27          second floor       red   

    Skin_Color Hair_Color  Height  
0       indigo     orange     191  
1       indigo      green     198  
2          red     violet     154  
3         blue     violet     158  
4       yellow     orange     184  
..         ...        ...     ...  
343     indigo        red     176  
344     indigo     orange     193  
345     indigo        red     193  
346     yellow     indigo     161  
347     yellow     orange     169  

[348 rows x 8 columns]

**TASK 2.2 (5 Points):** Now that we have cleaned up the Birthday column, lets find the age of each person. Create a new column called "Age" that displays the age of the person **on the day of the crime.** Once you are done, display the
"Profiles" dataframe

In [11]:
def calculateAge(birthday):
    today = pd.Timestamp.today().date()
    age = today.year - birthday.year - ((today.month, today.day) < (birthday.month, birthday.day))
    return age
    

profiles_df['Age'] = profiles_df['Birthday'].apply(calculateAge)
profiles_df

ID              Name   Birthday Location_during_crime Eye_Color  \
0    P299       Marco Rossi 1969-10-07          fourth floor       red   
1    P086       Aarav Patel 1979-10-03          second floor    indigo   
2    P022            Li Wei 1990-08-10          fourth floor      blue   
3    P003     Haruto Tanaka 1996-10-27                 lobby     green   
4    P036      Kwame Mensah 1991-02-26          fourth floor      blue   
..    ...               ...        ...                   ...       ...   
343  P126       Andrew Choe 1988-01-03                   NaN    violet   
344  P203     Krish Thakker 1965-09-26           third floor    yellow   
345  P273     Jeonghan Yoon 2003-07-12          fourth floor      blue   
346  P290  Ananya Malipeddi 1994-10-03           third floor       NaN   
347  P233         Ifra Syed 1982-06-27          second floor       red   

    Skin_Color Hair_Color  Height  Age  
0       indigo     orange     191   55  
1       indigo      green     198   45  
2          red     violet     154   34  
3         blue     violet     158   28  
4       yellow     orange     184   34  
..         ...        ...     ...  ...  
343     indigo        red     176   37  
344     indigo     orange     193   59  
345     indigo        red     193   21  
346     yellow     indigo     161   30  
347     yellow     orange     169   42  

[348 rows x 9 columns]

**TASK 2.3 (10 Points):** Now that we have created the "Age"; lets clean the 'Eye_color's column.

***Please read the directions carefully:***


- In order to fill in the missing eye_colors, we want to find the **most common eye_color** among those who have the **same hair and skin color**.
- For Instance, if we want to find the eye_color of someone who has yellow hair and violet skin:
 - we would first find everything who has the same hair and skin color, in this case it would be yellow hair and violet skin
 - and then we find the most common eye_color among them.
 - we then replace the missing eye_color with the eye color that is the most common among the people with the same hair and skin color.
- Watch out for redundant data!

- ***NOTE: In the case that you find multiple modes, choose the color that appears first in alphabetical order***

 When you finish filling in the missing eye colors, please **display the profiles_df**


In [12]:
def cleanEyeColor(person, df):
    if pd.isnull(person['Eye_Color']):
        # Get every eye color for those who match hair and skin color
        eye_colors = profiles_df[(profiles_df['Hair_Color'] == person['Hair_Color']) & 
                                 (profiles_df['Skin_Color'] == person['Skin_Color']) & 
                                 (profiles_df['Eye_Color'].notnull())]['Eye_Color'] 
        colors = Counter(eye_colors)
        max_count = max(colors.values())

        most_common_colors = []
        for color in colors.items():
            if color[1] == max_count:
                most_common_colors.append(color[0])

        # Must pick first appearing in alphabet. No effect if only one
        sorted_most_common_colors = sorted(most_common_colors)
        return sorted_most_common_colors[0]
    else:
        return person['Eye_Color']

profiles_df['Eye_Color'] = profiles_df.apply(lambda person: cleanEyeColor(person, profiles_df), axis=1)
profiles_df

ID              Name   Birthday Location_during_crime Eye_Color  \
0    P299       Marco Rossi 1969-10-07          fourth floor       red   
1    P086       Aarav Patel 1979-10-03          second floor    indigo   
2    P022            Li Wei 1990-08-10          fourth floor      blue   
3    P003     Haruto Tanaka 1996-10-27                 lobby     green   
4    P036      Kwame Mensah 1991-02-26          fourth floor      blue   
..    ...               ...        ...                   ...       ...   
343  P126       Andrew Choe 1988-01-03                   NaN    violet   
344  P203     Krish Thakker 1965-09-26           third floor    yellow   
345  P273     Jeonghan Yoon 2003-07-12          fourth floor      blue   
346  P290  Ananya Malipeddi 1994-10-03           third floor    yellow   
347  P233         Ifra Syed 1982-06-27          second floor       red   

    Skin_Color Hair_Color  Height  Age  
0       indigo     orange     191   55  
1       indigo      green     198   45  
2          red     violet     154   34  
3         blue     violet     158   28  
4       yellow     orange     184   34  
..         ...        ...     ...  ...  
343     indigo        red     176   37  
344     indigo     orange     193   59  
345     indigo        red     193   21  
346     yellow     indigo     161   30  
347     yellow     orange     169   42  

[348 rows x 9 columns]


## Q3) Finding the features of the culprit (38 PTS Total)

Now that we have successfully cleaned all of our data, lets begin finding the features of the criminal. Take a moment to look at the **"Witnesses"** table. We don't necessarily want all the statements from every single witness.

**NOTE:** Take a look at the structure of each statement, notice how there are only a few variations of statements from the witnesses.

**TASK 3.1 (5 Points):** Find all witnesses that were near the crime scene
- First load in the Witness.csv file and set it to a variable called "**witness_df**"
- create a new dataframe called **"important_witnesses"** that contains the id and the witness statements of those who were near the scene of the crime
- display the new dataframe when you are finished
- **HINT:** look at the floor which the crime happened on.

In [13]:
witness_df = pd.read_csv("Witness.csv")

# Get ID of everyone that has Location_during_crime as 'second floor'
important_people_ids = profiles_df[profiles_df["Location_during_crime"] == 'second floor']['ID']

important_witnesses = pd.merge(important_people_ids, witness_df, on="ID")
important_witnesses

ID                                          Statement
0   P086     I saw someone with red skin acting suspicious.
1   P157  I saw someone who looked 152 cm tall acting su...
2   P160   I saw someone with green skin acting suspicious.
3   P075  I saw someone with orange hair acting suspicious.
4   P224  I saw someone who looked 152 cm tall acting su...
..   ...                                                ...
66  P221    I saw someone with blue eyes acting suspicious.
67  P074    I saw someone with blue hair acting suspicious.
68  P174  I saw someone who looked 151 cm tall acting su...
69  P160   I saw someone with green skin acting suspicious.
70  P233  I saw someone that looks to be 22 years old ac...

[71 rows x 2 columns]

**TASK 3.2 (10 Points):** Find every potential eye color that the culprit could have

 - Put all the potential eye colors of the culprit in a list called **"potential_eye_colors"** and **display** it when you're done.
 - **potential_eye_colors** should have **no duplicates.**

In [14]:
potential_eye_colors = profiles_df['Eye_Color'].unique()
potential_eye_colors

array(['red', 'indigo', 'blue', 'green', 'orange', 'yellow', 'violet'],
      dtype=object)

**TASK 3.3 (5 Points):** Find every potential hair color that the culprit could have

 - Put all the potential hair colors of the culprit in a list called **"potential_hair_colors"** and **display** it when you're done.

In [15]:
potential_hair_colors = profiles_df['Hair_Color'].unique()
potential_hair_colors

array(['orange', 'green', 'violet', 'blue', 'red', 'indigo', 'yellow'],
      dtype=object)

**TASK 3.4 (5 Points):** Find every potential skin color that the culprit could have

 - Put all the potential skin colors of the culprit in a list called **"potential_skin_colors"** and display it when you're done

In [16]:
potential_skin_colors = profiles_df['Skin_Color'].unique()
potential_skin_colors

array(['indigo', 'red', 'blue', 'yellow', 'violet', 'orange', 'green'],
      dtype=object)

**TASK 3.5 (10 Points):** Find the potential height of the culprit

 - Put the smallest and largest potential heights of the culprit in the variables called **"smallest_height" & "largest_height"** and display them when you are done

In [17]:
smallest_height = profiles_df['Height'].min()
largest_height = profiles_df['Height'].max()
print("Smallest height: ", smallest_height, "; Largest height: ", largest_height)

Smallest height:  150 ; Largest height:  200


**TASK 3.6 (8 Points):** Find the potential age of the culprit

 - Put the smallest and largest potential age of the culprit in the variables called **"smallest_age" & "largest_age"** and display them when you are done

In [18]:
smallest_age = profiles_df['Age'].min()
largest_age = profiles_df['Age'].max()
print("Smallest age: ", smallest_age, "; Largest age: ", largest_age)

Smallest age:  18 ; Largest age:  65


## Q4) Who owns the Ticket? (30 PTS Total)

**TASK 4.1 (2 Points):** Loading final dfs


Load in the **"Zichao's Magic Show.csv"** into a dataframe called **"show_df"** and the **"Interrogation_Statement.csv"** into a dataframe called **"interrogation_df"** respectively and display them when you are done

In [19]:
show_df = pd.read_csv("Zichao's Magic Show.csv")
interrogation_df = pd.read_csv("Interrogation_Statements.csv")
print(show_df, interrogation_df)

       ID  Ticket_Number
0    P132       50269975
1    P056       28120732
2    P250       88051289
3    P197       79545169
4    P012       83521845
..    ...            ...
146  P121       31341288
147  P003       85008158
148  P180       24491860
149  P283       38524355
150  P142        2238676

[151 rows x 2 columns]                      Name                                          Statement
0    Kristopher Stoichkov  "Re%spe%cti%ng o%the%rs' be%lo%ngi%ngs i%s my ...
1           Naman Nagelia  "I@'d ne@ve@r je@o@pa@rdi@ze@ my re@pu@ta@ti@o...
2    Pranavh Vallabhaneni  "My pro^je^ct pre^se^nta^ti^o^n o^ccu^pi^e^d m...
3               Stacy Sun  "Thre$e$ co$nse$cu$ti$ve$ e$xa$ms dra$i$ne$d m...
4          Rina Kobayashi  "I% wa%s o%ff-ca%mpu%s a%tte%ndi%ng a% fa%mi%l...
..                    ...                                                ...
495             Ishan Kar  "I$ sto$le$ i$t i$n 2024 a$nd bri$be$d the$ po...
496          Adrian Black  "I% wa%s bu%sy wi%th my pro%je%ct

We're **so close** to figuring out the *last* and *final* clue to locate the culprit, but it looks like they are trying to beat us to it! What a devious criminal!

The culprit must have **sabotaged** our data after **interrogation** because it looks so wonky with weird symbols.

However, HQ informed us that our good friend and exemplar cryptanalyst **Alan Turing** has something to say to help us fix the data. **He encoded his message to keep it safe from the culprit.**

**TASK 4.2 (10 Points):**

Find what **Alan Turing** has to say:
- Use the **interrogation_df** to locate Alan's Turing's message.
- Put the message in a variable called "**encoded_message**" as a string (**Exclude** any **quotation marks** within the statement)
 - **Print** the encoded message in the following format: {"Encoded Message:", encoded_message}
- Figure out what the message says by taking a look at the numbers:
 - A **single space** between numbers indicates a letter to form a word.
 - A **double space** between numbers indicates the start of a new word.
- Store his decoded message in a variable called "**decoded_message**"
- Print out **Alan Turing's** decoded message in the following format: {"Decoded Message:", decoded_message}

**Hint:** Maybe [enumerating](https://docs.python.org/3/library/functions.html#enumerate) to store letters to numbers might help along with using [ascii values](https://docs.python.org/3/library/string.html)

In [20]:
encoded_message = interrogation_df[interrogation_df['Name'] == 'Alan Turing']['Statement'].item()
encoded_message = encoded_message.strip('"')
print("Encoded Message:", encoded_message)

# Create decoder
alphabet = string.ascii_lowercase
decoder = sorted(list(enumerate(alphabet, start=1)))

# Split statement into words
words = encoded_message.split("  ")

decoded_message = ""
#loop through words
for word in words:
    letters = word.split(" ")
    for letter in letters:
        integer, decoded_letter = decoder[int(letter) - 1]
        decoded_message += decoded_letter

    # add space between words
    decoded_message += " "

print("Decoded Message:", decoded_message)

Encoded Message: 20 8 5  16 1 20 20 5 18 14  9 19  20 15  3 12 5 1 14  21 16  20 8 5  6 9 12 5  2 25  18 5 13 15 22 9 14 7  1 6 20 5 18  5 1 3 8  22 15 23 5 12  20 8 5  19 25 13 2 15 12 19  16 5 18 3 5 14 20  4 15 12 12 1 18  3 1 18 18 15 20  1 20
Decoded Message: the pattern is to clean up the file by removing after each vowel the symbols percent dollar carrot at 


**TASK 4.3 (10 Points):**

**Restore** the interrogation statements back to their original state after **following the decoded message from before**.
- After restoration, make sure your modifications are updated in the **interrogation_df** itself.
- **Display** the **interrogation_df** when you are done.

In [21]:
def restoreStatments(statement):
    restored = statement.replace('%', '').replace("@", '').replace('$', '').replace("^", '')
    return restored

interrogation_df['Statement'] = interrogation_df['Statement'].apply(restoreStatments)
interrogation_df

Name                                          Statement
0    Kristopher Stoichkov  "Respecting others' belongings is my principle...
1           Naman Nagelia  "I'd never jeopardize my reputation by stealin...
2    Pranavh Vallabhaneni  "My project presentation occupied me in the en...
3               Stacy Sun  "Three consecutive exams drained me that day. ...
4          Rina Kobayashi  "I was off-campus attending a family event. Se...
..                    ...                                                ...
495             Ishan Kar  "I stole it in 2024 and bribed the police with...
496          Adrian Black  "I was busy with my project presentation in an...
497       Brittany Taylor  "I was busy with my group project meeting in a...
498        William Miller  "I wasn't even on campus that day. Check the a...
499  Christopher Ferguson  "Stealing? That's not in my nature at all. I r...

[500 rows x 2 columns]


**TASK 4.4 (8 Points):** Create a column called **Ticket_Color** for the **"show_df"** that displays all the colors of the ticket

Remember that there was a show ticket left at the crime scene, take a second to look through **"show_df"**.
- Try and find a way to identify how you can find the ticket color from the ticket number.
- **Display** the **"show_df"** when you are done

**HINT:** Examine the **"interrogation_df"**, someone might have an idea

In [22]:
# Monkey D Luffy,"""Sum the individual #s in ticket numbers, look at the rightmost value, 0-2:B, 3-5:R , 6-7:G, 8-9:Y"""
# from looking at statements with 'ticket' in it
def ticketColor(number):
    total = 0
    for individual_number in str(number):
        total += int(individual_number)

    rightmost_number = int(str(total)[-1])
    if rightmost_number in [0,1,2]:
        return 'Blue'
    elif rightmost_number in [3,4,5]:
        return 'Red'
    elif rightmost_number in [6,7]:
        return 'Green'
    else:
        return 'Yellow'

show_df['Ticket_Color'] = show_df['Ticket_Number'].apply(ticketColor)
show_df

ID  Ticket_Number Ticket_Color
0    P132       50269975          Red
1    P056       28120732          Red
2    P250       88051289         Blue
3    P197       79545169        Green
4    P012       83521845        Green
..    ...            ...          ...
146  P121       31341288         Blue
147  P003       85008158          Red
148  P180       24491860          Red
149  P283       38524355          Red
150  P142        2238676          Red

[151 rows x 3 columns]

## Q5) Finding the culprit (12 PTS Total)

A helpful function might be the "iloc" function. Have a look at: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html

**TASK 5.1(10 Points):** Using all the **data compiled from before** find a person that matches all the features from the profile.

- Do not hardcode or you will lose credit after all your hardwork 🥸

- Put the name of the culprit in a variable called **"culprit"**
- Print out the name of the culprit (There is only one).

In [69]:
# look at important_witnesses and see if any had a red ticket to the show
red_tickets = show_df[show_df['Ticket_Color'] == 'Red']['ID']

red_important_witness = pd.merge(red_tickets, important_witnesses, on='ID')

red_second = pd.merge(red_tickets, important_witnesses, on='ID')
red_second_no_duplicates = red_second.drop_duplicates(subset='ID', keep=False)


# from important winesses see most common eye color
eyes = []
for statement in important_witnesses['Statement']:
    words = statement.split()
    for i in range(1, len(words)):
        if words[i] == 'eyes':
            eyes.append(words[i-1])


eye_mode = statistics.mode(eyes)

profile = profiles_df[(profiles_df['ID'].isin(red_second_no_duplicates['ID'])) & (profiles_df['Eye_Color']==eye_mode)]
culprit = profile['Name']
culprit




283    LeBron James
Name: Name, dtype: object

**TASK 5.2 (2 Points):** Check if your culprit is correct!

**Interrogate** your culprit to see if you got the correct person.

In [35]:
print(interrogation_df[interrogation_df['Name']==culprit]['Statement'])

468    "Forgive me sunshine. The markers were just to...
Name: Statement, dtype: object


# GREAT WORK DETECTIVE!! 🌻